server/cgmon.go

```diff
@@ -0,0 +1,135 @@
+package server
+
+import (
+	"context"
+	"fmt"
+	"math"
+	"runtime"
+	"sync"
+	"time"
+
+	"github.com/pingcap/log"
+	"github.com/shirou/gopsutil/v3/mem"
+	bs "github.com/tikv/pd/pkg/basicserver"
+	"github.com/tikv/pd/pkg/cgroup"
+	"go.uber.org/zap"
+)
+
+const (
+	refreshInterval = 10 * time.Second
+)
+
+type cgroupMonitor struct {
+	started         bool
+	ctx             context.Context
+	cancel          context.CancelFunc
+	wg              sync.WaitGroup
+	cfgMaxProcs     int
+	lastMaxProcs    int
+	lastMemoryLimit uint64
+}
+
+// StartCgroupMonitor uses to start the cgroup monitoring.
+// WARN: this function is not thread-safe.
+func (cgmon *cgroupMonitor) startCgroupMonitor() {
+	if cgmon.started {
```

Comment:
Will it be started twice?

Reply:
> Will it be started twice?
No. Just to make sure it won't. In fact we originally implemented cgroup monitor in tidb and the started flag stuff referred to the logic of background routines like the topsql cpu collector.